internal/model/application.go

```diff
@@ -0,0 +1,69 @@
+package model
+
+import (
+	"github.com/G-Research/yunikorn-core/pkg/webservice/dao"
+)
+
+type Application struct {
+	ModelMetadata          `json:",inline"`
+	dao.ApplicationDAOInfo `json:",inline"`
+}
+
+func (app *Application) MergeFrom(appInfo *dao.ApplicationDAOInfo) {
+	app.Partition = appInfo.Partition
+	app.QueueName = appInfo.QueueName
+	app.SubmissionTime = appInfo.SubmissionTime
+	app.FinishedTime = appInfo.FinishedTime
+	app.State = appInfo.State
+	app.User = appInfo.User
+	app.Groups = appInfo.Groups
+	app.RejectedMessage = appInfo.RejectedMessage
+	app.StateLog = appInfo.StateLog
+	app.HasReserved = appInfo.HasReserved
+	app.Reservations = appInfo.Reservations
+	app.MaxRequestPriority = appInfo.MaxRequestPriority
+	app.StartTime = appInfo.StartTime
+	app.PlaceholderData = appInfo.PlaceholderData
+
+	lookup := make(map[string]struct{})
+	if len(appInfo.Allocations) > 0 {
+		for _, alloc := range app.Allocations {
+			lookup[alloc.AllocationKey] = struct{}{}
+		}
+		for _, alloc := range appInfo.Allocations {
+			if _, ok := lookup[alloc.AllocationKey]; !ok {
+				app.Allocations = append(app.Allocations, alloc)
+			}
+		}
+		clear(lookup)
+	}
+	if len(appInfo.Requests) > 0 {
+		for _, ask := range app.Requests {
+			lookup[ask.AllocationKey] = struct{}{}
+		}
+		for _, ask := range app.Requests {
+			if _, ok := lookup[ask.AllocationKey]; !ok {
+				app.Requests = append(app.Requests, ask)
+			}
+		}
+		clear(lookup)
+	}
+	if appInfo.UsedResource != nil {
```

Comment:
What is the difference than just doing:
```
app.UsedResource = appInfo.UsedResource
```

I think the if is redundant, if it is nil, it will also be nil in `app`.

Reply:
Doesn't have to be
App accumulates stuff during many events, so if during deletion, used resources is set to nil, I want to preserve the old data. App is persisted in the database, and later merged on event.